video/updater.h

```diff
@@ -0,0 +1,160 @@
+#ifndef UPDATER_H
+#define UPDATER_H
+
+#include "vdu_stream_processor.h"
+#include <cstdint>
+#include "esp_ota_ops.h"
+
+extern void printFmt(const char *format, ...);
+extern void print(const char * text);
+
+namespace Updater {
+	static constexpr uint8_t unlockCode[] = "unlock";
+	static bool isLocked = true;
+}
+
+void VDUStreamProcessor::unlock() {
+	uint8_t buffer[sizeof(Updater::unlockCode)] = {0};
+	uint32_t bytes_remain = readIntoBuffer(buffer, sizeof(buffer) - 1);
+	if(bytes_remain) {
+		printFmt("Read unlock code failed!\n\r");
+		return;
+	}
+	else if(memcmp(buffer, Updater::unlockCode, sizeof(Updater::unlockCode)) == 0) {
+		printFmt("Updater unlocked!\n\r");
+		Updater::isLocked = false;
+		return;
+	}
+}
+
+void VDUStreamProcessor::receiveFirmware() {
+	uint32_t update_size = 0;
+	uint32_t bytes_remain = readIntoBuffer((uint8_t*)&update_size, sizeof(update_size) - 1); // -1 because its 24bits
+	if(bytes_remain) {
+		printFmt("Read size failed!\n\r");
+		return;
+	}
+
+	if(Updater::isLocked) {
+		printFmt("Updater is locked, bytes will be discarded!\n\r");
+		discardBytes(update_size + 1);
+		return;
+	}
+	
+	printFmt("Received update size: %u bytes\n\r", update_size);
+
+	printFmt("Receiving VDP firmware update");
+
+	uint32_t start = millis();
+
+	esp_err_t err;
+	esp_ota_handle_t update_handle = 0 ;
+	const esp_partition_t *update_partition = NULL;
+	const esp_partition_t *configured = esp_ota_get_boot_partition();
+	const esp_partition_t *running = esp_ota_get_running_partition();
+
+	update_partition = esp_ota_get_next_update_partition(NULL);
+	err = esp_ota_begin(update_partition, OTA_SIZE_UNKNOWN, &update_handle);
+	if (err != ESP_OK) {
+		printFmt("esp_ota_begin failed, error=%d\n\r", err);
```

Comment:
if I've understood correctly, this is a failure condition and if it happens the update won't be able to proceed.  therefore I think this should become:

```suggestion
		printFmt("esp_ota_begin failed, error=%d\n\r", err);
		discardBytes(update_size + 1);
		return;
```